### ISSUE TITLE
Random query failures with AllegroGraph v7.0.1

### ISSUE THREAD

**Issue by mdorf** [ncbo/goo#107]
The query that fails is always the same (different graphs and ID filters, but the structure is identical), but the failures happen at random in various tests during the test suite execution. In each subsequent run, the error occurs inside a different test, where as the test that had previously failed runs fine. 

This is the error:
```
Executing query failed:
the assertion (null (getf (db.agraph.cursor::cursor-properties db.agraph.sbqe::cursor) :is-ordered-cursor-for-distinct-variable)) failed.
```

This is the query:
```
SELECT DISTINCT ?id ( COUNT(DISTINCT ?children_agg_count) AS ?children_agg_count_projection ) 
FROM <http://data.bioontology.org/ontologies/OBSBRANCH/submissions/1> 
WHERE { ?id a <http://www.w3.org/2002/07/owl#Class> . 
       OPTIONAL { ?children_agg_count <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?id .  } 
       FILTER(?id = <http://www.w3.org/2004/02/skos/core#oboClass2OBS> || 
              ?id = <http://www.w3.org/2004/02/skos/core#oboClass1OBS> || 
              ?id = <http://www.geneontology.org/formats/oboInOwl#ObsoleteClass> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class_7> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class_5> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class99> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class98> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class97> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class6> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class4> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class3> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class2> || 
              ?id = <http://bioportal.bioontology.org/ontologies/msotes#class1>) } 
GROUP BY ?id
```

This does NOT occur in AllegroGraph v6.4.1

**Comment by mdorf** [comment]
Further response from AllegroGraph:

> Thanks for the report and the query log. We have traced it down and the next AG release, 7.0.2, will have a fix for this.
> 
> The reason the error does not occur every time you run the same SPARQL query, is that there are multiple possible query plans and only some of them lead to the bug. Which plan gets chosen depends on statistics (number of triples per predicate, and such).
> 
> As a workaround for now I recommend you create a patch file with the following name and contents:
> 
> filename: <AG directory> /lib/patches/patch-spr45071-bug26235.cl
> contents:
> 
>  (in-package :db.agraph.cursor)
> 
>  (excl:def-fwrapper make-null-spog-cursor-wrapper
>     (&key order-by extended-triples-p)
>  (declare (ignore order-by extended-triples-p))
>  (let ((ret (excl:call-next-fwrapper)))
>    (setf (cursor-properties ret) nil)
>    ret))
> 
>  (excl:fwrap 'make-null-spog-cursor
>              'make-null-spog-cursor-wrapper
>              'make-null-spog-cursor-wrapper)
> 
>  (compile 'make-null-spog-cursor-wrapper)
> 
> 
> After creating this file and restarting AG the query should run fine all the time. Please remove this patch file when updating to AG 7.0.2. Thanks again for the report.
>

**Comment by mdorf** [comment]
I've confirmed that the patch fixes the issue. Ran the tests multiple times with no failures

**Comment by mdorf** [comment]
Currently testing AllegroGraph v7.0.2.rc2.

**Comment by mdorf** [comment]
fixed in v7.0.2
